processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java

```diff
@@ -97,8 +106,10 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
     final TopNAlgorithmSelector selector = new TopNAlgorithmSelector(cardinality, numBytesPerRecord);
     query.initTopNAlgorithmSelector(selector);
 
-    TopNAlgorithm topNAlgorithm = null;
-    if (selector.isHasDimExtractionFn()) {
+    final TopNAlgorithm topNAlgorithm;
+    if (selector.isHasExtractionFn() && dimension.equals(Column.TIME_COLUMN_NAME)) {
```

Comment:
Agree, they're not pretty. In this case `dimension.equals(Column.TIME_COLUMN_NAME)` really means "is this a dimension of type long and of unknown cardinality". I'm not sure if there is a better way to express that in the current state, but I can add a comment.